sympy/matrices/decompositions.py

```diff
@@ -1343,3 +1344,72 @@ def _QRdecomposition(M):
     QRsolve
     """
     return _QRdecomposition_optional(M, normalize=True)
+
+def _upper_hessenberg_decomposition(A):
+    """Converts a matrix into Hessenberg matrix H
+
+    Returns 2 matrices H, P s.t.
+    $P H P^{T} = A$, where H is an upper hessenberg matrix
+    and P is an orthogonal matrix
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> A = Matrix([
+    ...     [1,2,3],
+    ...     [-3,5,6],
+    ...     [4,-8,9],
+    ... ])
+    >>> H, P = A.upper_hessenberg_decomposition()
+    >>> H
+    Matrix([
+    [1,    6/5,    17/5],
+    [5, 213/25, -134/25],
+    [0, 216/25,  137/25]])
+    >>> P
+    Matrix([
+    [1,    0,   0],
+    [0, -3/5, 4/5],
+    [0,  4/5, 3/5]])
+    >>> P * H * P.H == A
```

Comment:
You also need to consider upgrading this method to work with complex matrix with complex householder transformation. Using `.H` for methods that work only with real matrices can be an abuse of notation.

Reply:
```python
A = Matrix([
    [1, 1 + 2*I, 2],
    [1 - 3*I, -1, 3],
    [1 + 3*I, 1, 0]
])
H, P = A.upper_hessenberg_decomposition()
pprint(simplify(H))
```

    ⎡                                                        3⋅√2   8⋅√2⋅ⅈ        
    ⎢     1                                                  ──── - ──────        
    ⎢                                                         10      5           
    ⎢                                                                             
    ⎢              ⅈ⋅(-40⋅√5⋅(3 + 20⋅ⅈ) - 12⋅√10⋅(7 + 47⋅ⅈ) + 20⋅√5⋅(3 + 16⋅ⅈ)⋅(√2
    ⎢-√2 + 3⋅√2⋅ⅈ  ───────────────────────────────────────────────────────────────
    ⎢                                                  10⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10) 
    ⎢                                                                             
    ⎢                   (4 - 3⋅ⅈ)⋅(146⋅√5⋅(√2 + 2) + 107⋅√10⋅(√2 + 2) - 10⋅√10⋅(23
    ⎢     0             ──────────────────────────────────────────────────────────
    ⎣                                                  50⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10) 
    
                                                                                  
                                                                                  
                                                                                  
                                                                                  
     + 2) + √10⋅(42 + 227⋅ⅈ)⋅(√2 + 2))  (4 + 3⋅ⅈ)⋅(√10⋅(155 - 72⋅ⅈ)⋅(√2 + 2) + 6⋅√
    ──────────────────────────────────  ──────────────────────────────────────────
                                                                                  
                                                                                  
     + 12⋅ⅈ) - 12⋅√5⋅(27 + 14⋅ⅈ))               ⅈ⋅(-20⋅√10⋅(3 + 8⋅ⅈ) - 28⋅√5⋅(3 + 
    ─────────────────────────────               ──────────────────────────────────
                                                                                 1
    
         4⋅√2   11⋅√2⋅ⅈ                                                ⎤
         ──── + ───────                                                ⎥
          5        10                                                  ⎥
                                                                       ⎥
    5⋅(35 - 16⋅ⅈ)⋅(√2 + 2) + 4⋅√5⋅(-193 + 126⋅ⅈ) + 10⋅√10⋅(-55 + 36⋅ⅈ))⎥
    ───────────────────────────────────────────────────────────────────⎥
    50⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10)                                         ⎥
                                                                       ⎥
    8⋅ⅈ) + √5⋅(6 + 31⋅ⅈ)⋅(√2 + 2) + 2⋅√10⋅(3 + 13⋅ⅈ)⋅(√2 + 2))         ⎥
    ──────────────────────────────────────────────────────────         ⎥
    0⋅(2⋅√10 + 3⋅√5)⋅(√2 + 2)                                          ⎦